### PR TITLE
Issue 809:  Implement a PypiWheelArtifactCoordinates

### LINKED ISSUE #809: Implement a PypiWheelArtifactCoordinates
# Task Description

We need to identify what the artifact coordinates are for PyPi packages and implement a class for this. 

The Python Wheel artifact coordinates are described [here](https://www.python.org/dev/peps/pep-0427/#file-format). They can be inferred from the path.

## Specification

*File name convention*

The wheel filename is `{distribution}-{version}(-{build tag})?-{python tag}-{abi tag}-{platform tag}.whl`.

* `distribution`
Distribution name, e.g. 'django', 'pyramid'.
* `version`
Distribution version, e.g. `1.0`.
* `build tag`
Optional build number. Must start with a digit. A tie breaker if two wheels have the same version. Sort as the empty string if unspecified, else sort the initial digits as a number, and the remainder lexicographically.
* `language implementation and version tag`
E.g. `py27`, `py2`, `py3`.
* `abi tag`
E.g. `cp33m`, `abi3`, `none`.
* `platform tag`
E.g. `linux_x86_64`, `any`.
For example, `distribution-1.0-1-py27-none-any.whl` is the first build of a package called `distribution`, and is compatible with Python `2.7` (any Python 2.7 implementation), with no ABI (pure Python), on any CPU architecture.

The last three components of the filename before the extension are called "compatibility tags." The compatibility tags express the package's basic interpreter requirements and are detailed in PEP 425.

# Where to place the code

The code should be located in a class called `PypiWheelArtifactCoordinates` under `strongbox/strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/main/java/org/carlspring/strongbox/artifact/coordinates`.

Have a look at the [Artifact Coordinates](https://strongbox.github.io/developer-guide/layout-providers/how-to-implement-your-own-repository-format.html#artifact-coordinates) wiki page for more information.

# Tests

We will need tests to be implemented for this that cover:
* Parsing a list of real-world package names and making sure their coordinates work
* Parsing of artifact coordinates where some of the option parameters have not been specified

# Task Relationships

This is a sub-task of #807. It should be relatively straight-forward to implement this and you should get acquainted with the master task, before proceeding.

# Help

* [Our chat](https://chat.carlspring.org/)
* Points of contact:
  * @carlspring
  * @sbespalov
  * @fuss86

### PR DESCRIPTION
**Pull request by Alecg956**
- Implemented PypiWheelArtifactCoordinates.java
- Implemented PypiWheelArtifactCoordinatesTest.java
- Implemented PypiWheelArtifactCoordinatesUtils.java

Fixes #809.

### CODE REVIEW

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/main/java/org/carlspring/strongbox/artifact/coordinates/PypiWheelArtifactCoordinates.java [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+     */
+    public PypiWheelArtifactCoordinates(String distribution,
+                                        String version,
+                                        String build,
+                                        String languageImplementationVersion,
+                                        String abi,
+                                        String platform)
+    {
+        // if any of the required arguments are empty, throw an error
+        if (distribution == "" || version == "" || languageImplementationVersion == "" || abi == "" || platform == "")
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/main/java/org/carlspring/strongbox/artifact/coordinates/PypiWheelArtifactCoordinates.java]: In Java, comparing objects should be done using `equals()`. Checking using `==` checks that two operands point to the same object. Have a look at these articles:
* https://stackoverflow.com/a/13387787/774183
* https://www.leepoint.net/data/expressions/22compareobjects.html
* https://www.geeksforgeeks.org/equals-hashcode-methods-java/

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/main/java/org/carlspring/strongbox/artifact/coordinates/PypiWheelArtifactCoordinates.java [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+import javax.xml.bind.annotation.XmlAccessorType;
+import javax.xml.bind.annotation.XmlRootElement;
+import java.util.Map;
+
+import org.carlspring.strongbox.util.PypiWheelArtifactCoordinatesUtils;
+import org.semver.Version;
+import org.apache.commons.lang3.StringUtils;
+
+/**
+ * Represents {@link ArtifactCoordinates} for PyPi repository
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/main/java/org/carlspring/strongbox/artifact/coordinates/PypiWheelArtifactCoordinates.java]: Shouldn't this be `PypiWheelArtifactCoordinates`?

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/main/java/org/carlspring/strongbox/artifact/coordinates/PypiWheelArtifactCoordinates.java [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+@XmlAccessorType(XmlAccessType.NONE)
+@ArtifactCoordinatesLayout(name = PypiWheelArtifactCoordinates.LAYOUT_NAME, alias = PypiWheelArtifactCoordinates.LAYOUT_ALIAS)
+public class PypiWheelArtifactCoordinates
+    extends AbstractArtifactCoordinates<PypiWheelArtifactCoordinates, Version>
+{
+
+    public static final String LAYOUT_NAME = "PyPi";
+    public static final String LAYOUT_ALIAS = "pypi";
+
+    public static final String DISTRIBUTION = "distribution";
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/main/java/org/carlspring/strongbox/artifact/coordinates/PypiWheelArtifactCoordinates.java]: Would you mind changind these variables as follows:
```
public static final String BUILD = "build";

public static final String LANGUAGE_IMPLEMENTATION_VERSION = "languageImplementationVersion";

private static final String ABI = "abi";

private static final String PLATFORM = "platform";
```

Cosmetic: Could you please keep an empty line between fields and constants?

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/main/java/org/carlspring/strongbox/artifact/coordinates/PypiWheelArtifactCoordinates.java [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+     * This method takes in all artifact coordinates of a PyPi Wheel filename, with build being 
+     * the empty string if it is not included in the filename
+     * 
+     * @param distribution: Uniquely identifying artifact coordinate (required)
+     * @param version: Packages current version (required)
+     * @param build: Build_tag parameter (optional)
+     * @param languageImplementationVersion: Language and Implementation version argument (required)
+     * @param abi: ABI tag parameter (required)
+     * @param platform: Platform tag parameter (required)
+     * @return bar Returns bar
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/main/java/org/carlspring/strongbox/artifact/coordinates/PypiWheelArtifactCoordinates.java]: This is a constructor. It doesn't return anything, so you can drop this.

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/main/java/org/carlspring/strongbox/artifact/coordinates/PypiWheelArtifactCoordinates.java [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+    public static final String BUILD_TAG = "build_tag";
+    public static final String LANG_IMPL_VERSION_TAG = "languageImplementationVersion_tag";
+    private static final String ABI_TAG = "abi_tag";
+    private static final String PLATFORM_TAG = "platform_tag";
+
+    /**
+     * This method takes in all artifact coordinates of a PyPi Wheel filename, with build being 
+     * the empty string if it is not included in the filename
+     * 
+     * @param distribution: Uniquely identifying artifact coordinate (required)
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/main/java/org/carlspring/strongbox/artifact/coordinates/PypiWheelArtifactCoordinates.java]: In Javadocs, the format is something like this:
```
 * @param   version   Packages current version (required)
```

There are shouldn't be a `:` between the varibale and the description.

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/main/java/org/carlspring/strongbox/artifact/coordinates/PypiWheelArtifactCoordinates.java [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+                                        String abi,
+                                        String platform)
+    {
+        // if any of the required arguments are empty, throw an error
+        if (distribution == "" || version == "" || languageImplementationVersion == "" || abi == "" || platform == "")
+        {
+            throw new IllegalArgumentException("The distribution, version, language_implementation_version_tag, abi_tag, and platform_tag are mandatory fields.");
+        }
+
+        if (build != "" && !Character.isDigit(build.charAt(0)))
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/main/java/org/carlspring/strongbox/artifact/coordinates/PypiWheelArtifactCoordinates.java]:
```
if (!StringUtils.isBlank(build) && !Character.isDigit(build.charAt(0)))
```

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/main/java/org/carlspring/strongbox/artifact/coordinates/PypiWheelArtifactCoordinates.java [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+    extends AbstractArtifactCoordinates<PypiWheelArtifactCoordinates, Version>
+{
+
+    public static final String LAYOUT_NAME = "PyPi";
+    public static final String LAYOUT_ALIAS = "pypi";
+
+    public static final String DISTRIBUTION = "distribution";
+    public static final String VERSION = "version";
+    public static final String BUILD_TAG = "build_tag";
+    public static final String LANG_IMPL_VERSION_TAG = "languageImplementationVersion_tag";
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/main/java/org/carlspring/strongbox/artifact/coordinates/PypiWheelArtifactCoordinates.java]: The variable naming convention that we typically follow looks like this:
```
InputStream inputStream = ...;
```
Or:
```
InputStream is = ...;
```

Similarly, our fields and constants should look like this as well.

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/main/java/org/carlspring/strongbox/artifact/coordinates/PypiWheelArtifactCoordinates.java [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+    private static final String ABI_TAG = "abi_tag";
+    private static final String PLATFORM_TAG = "platform_tag";
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/main/java/org/carlspring/strongbox/artifact/coordinates/PypiWheelArtifactCoordinates.java]: I'm not quite sure why part of the coordinates are `private static final` whereas the others are `public static final`... I believe they should all be the latter.

**Comment by carlspring** [comment]
Hi @Alecg956 ,

Thanks for your contribution and congratulations on your first pull request! The code looks good! Could you please have a look at my concerns and address them? They're mosly trivial things to sort out!

Thanks! :)

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/main/java/org/carlspring/strongbox/artifact/coordinates/PypiWheelArtifactCoordinates.java (hunk 1 above)
Comment (Alecg956) [reply strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/main/java/org/carlspring/strongbox/artifact/coordinates/PypiWheelArtifactCoordinates.java]: Resolved.

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/main/java/org/carlspring/strongbox/artifact/coordinates/PypiWheelArtifactCoordinates.java (hunk 2 above)
Comment (Alecg956) [reply strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/main/java/org/carlspring/strongbox/artifact/coordinates/PypiWheelArtifactCoordinates.java]: From npm artifact coordinates:  " This class is an {@link ArtifactCoordinates} implementation for npm"

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/main/java/org/carlspring/strongbox/artifact/coordinates/PypiWheelArtifactCoordinates.java (hunk 3 above)
Comment (Alecg956) [reply strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/main/java/org/carlspring/strongbox/artifact/coordinates/PypiWheelArtifactCoordinates.java]: Should be resolved.

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/main/java/org/carlspring/strongbox/artifact/coordinates/PypiWheelArtifactCoordinates.java (hunk 4 above)
Comment (Alecg956) [reply strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/main/java/org/carlspring/strongbox/artifact/coordinates/PypiWheelArtifactCoordinates.java]: Removed.

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/main/java/org/carlspring/strongbox/artifact/coordinates/PypiWheelArtifactCoordinates.java (hunk 5 above)
Comment (Alecg956) [reply strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/main/java/org/carlspring/strongbox/artifact/coordinates/PypiWheelArtifactCoordinates.java]: Fixed.

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/main/java/org/carlspring/strongbox/artifact/coordinates/PypiWheelArtifactCoordinates.java (hunk 6 above)
Comment (Alecg956) [reply strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/main/java/org/carlspring/strongbox/artifact/coordinates/PypiWheelArtifactCoordinates.java]: Changed!

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/main/java/org/carlspring/strongbox/artifact/coordinates/PypiWheelArtifactCoordinates.java (hunk 7 above)
Comment (Alecg956) [reply strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/main/java/org/carlspring/strongbox/artifact/coordinates/PypiWheelArtifactCoordinates.java]: Changed.

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/main/java/org/carlspring/strongbox/artifact/coordinates/PypiWheelArtifactCoordinates.java (hunk 8 above)
Comment (Alecg956) [reply strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/main/java/org/carlspring/strongbox/artifact/coordinates/PypiWheelArtifactCoordinates.java]: I'm not sure either, will change them to all be public.

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/main/java/org/carlspring/strongbox/artifact/coordinates/PypiWheelArtifactCoordinates.java (hunk 2 above)
Comment (carlspring) [reply strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/main/java/org/carlspring/strongbox/artifact/coordinates/PypiWheelArtifactCoordinates.java]: Oh, yes, sorry, you're right!

**Comment by carlspring** [comment]
Thanks for polishing this up and for doing the research, development and testing!
This will be of great help for the implementation of the PyPi layout provider.

**Pull request merged**